src/bluez/adapter/mod.rs

```diff
@@ -11,517 +11,478 @@
-    Unknown(u8),
-}
+assert_impl_all!(SyncConnection: Sync, Send);
+assert_impl_all!(Adapter: Sync, Send);
 
-impl AdapterType {
-    fn parse(typ: u8) -> AdapterType {
-        match typ {
-            0 => AdapterType::BrEdr,
-            1 => AdapterType::Amp,
-            x => AdapterType::Unknown(x),
-        }
-    }
+impl Adapter {
+    pub(crate) fn from_dbus_path(path: &Path) -> Result<Adapter> {
+        let conn = Arc::new(SyncConnection::new_system()?);
+        let proxy = conn.with_proxy(BLUEZ_DEST, path, DEFAULT_TIMEOUT);
+        info!("DevInfo: {:?}", proxy.address()?);
+
+        let adapter = Adapter {
+            connection: conn,
+            listener: Arc::new(ReentrantMutex::new(SyncConnection::new_system()?)),
+            path: path.to_string(),
+            manager: AdapterManager::new(),
+            match_tokens: Arc::new(DashMap::new()),
+            should_stop: Arc::new(AtomicBool::new(false)),
+            thread_handle: Arc::new(Mutex::new(None)),
+        };
 
-    fn num(&self) -> u8 {
-        match *self {
-            AdapterType::BrEdr => 0,
-            AdapterType::Amp => 1,
-            AdapterType::Unknown(x) => x,
-        }
-    }
-}
+        adapter.setup();
 
-#[derive(Hash, Eq, PartialEq, Debug, Copy, Clone)]
-pub enum AdapterState {
-    Up,
-    Init,
-    Running,
-    Raw,
-    PScan,
-    IScan,
-    Inquiry,
-    Auth,
-    Encrypt,
-}
+        Ok(adapter)
+    }
 
-impl AdapterState {
-    fn parse(flags: u32) -> HashSet<AdapterState> {
-        use self::AdapterState::*;
+    fn setup(&self) {
+        use dbus::blocking::stdintf::org_freedesktop_dbus::ObjectManagerInterfacesRemoved as InterfacesRemoved;
+        let mut lost_rule = InterfacesRemoved::match_rule(None, None);
+        lost_rule.path = Some(Path::from("/"));
 
-        let states = [Up, Init, Running, Raw, PScan, IScan, Inquiry, Auth, Encrypt];
+        // // Enable watching the DBus channel for new messages
+        // conn.channel().set_watch_enabled(true);
+        // let watcher = conn.channel().watch();
+        // // watcher.
+        let listener = self.listener.clone();
+        let should_stop = self.should_stop.clone();
 
-        let mut set = HashSet::new();
-        for (i, f) in states.iter().enumerate() {
-            if flags & (1 << (i & 31)) as u32 != 0 {
-                set.insert(f.clone());
+        // Spawn a new thread to process incoming DBus messages.
+        // This thread gets cleaned up in the 'impl Drop for Adapter'
+        *self.thread_handle.lock().unwrap() = Some(thread::spawn(move || {
+            while !should_stop.load(Ordering::Relaxed) {
+                // listener is protected by a mutex. as calling `process()` when a proxy is awaiting a reply will cause the reply to be dropped...
+                let lock = listener.lock();
+                while lock.process(Duration::from_secs(0)).unwrap() {}
+                drop(lock);
+                thread::sleep(Duration::from_millis(100));
```

Comment:
no source for this. I was going for a way to quickly process any messages there may be, and then sleeping on the thread for 100ms, so that other threads have a chance to use the listener to send/receive

Reply:
Looking back at my code and later learning about `parking_lot` I was dealing with mutex _fairness_, where, the `while` loop was constantly locking the mutex, not giving any other threads a chance to get a lock, and thus, anything that needed `listener` would block. 

I'll see if `parking_lot::ReentrantMutex` promotes fairness